presto-expressions/src/main/java/com/facebook/presto/expressions/translator/RowExpressionTranslator.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.expressions.translator;
+
+import com.facebook.presto.spi.function.FunctionHandle;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+
+import java.util.List;
+import java.util.Optional;
+
+public class RowExpressionTranslator<T, C>
+{
+    /**
+     * translator for function that is recursive. Will be overidden if non-recursive version returns translated values.
+     */
+    public Optional<T> translateCall(FunctionHandle functionHandle, List<TranslatedExpression<T>> translatedExpressions, C context)
```

Comment:
All these calls should return `TranslatedExpression<T>`. It is up to TranslatedExpression to decide if they are optional empty or not.
Also, why we have two translateCalls and translateSpecialForms? We don't really need the ones handling `functionHandle` or  `SpecialFormExpression.Form`

Reply:
Yeah I agree. That makes sense, refactored as such.